File: src/main/java/com/flanks255/simplybackpacks/gui/FilterGui.java
package com.flanks255.simplybackpacks.gui;

import com.flanks255.simplybackpacks.SimplyBackpacks;
import com.mojang.blaze3d.platform.Lighting;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.item.ItemStack;

import javax.annotation.Nonnull;

public class FilterGui extends AbstractContainerScreen<FilterContainer> {
    public FilterGui(FilterContainer container, Inventory playerInventory, Component name) {
        super(container, playerInventory, name);

        inventory = playerInventory;
        this.imageWidth = 176;
        this.imageHeight = 166;
    }

    private final Inventory inventory;

    @Override
    protected void init() {
        super.init();

        Button.OnPress slotClick = button -> {
            Minecraft.getInstance().gameMode.handleInventoryButtonClick(menu.containerId, ((SlotButton)button).slot);
            menu.clickMenuButton(inventory.player, ((SlotButton)button).slot);
        };

        int slot = 0;
        for (int row = 0; row < 4; row ++) {
            for (int col = 0; col < 4; col++) {
                int x = this.leftPos + 7 + col * 18;
                int y = this.topPos + 7 + row * 18;

                addRenderableWidget(new SlotButton(x+1, y+1,18 ,18, slot, slotClick));
                slot++;
            }
        }

        addRenderableWidget(new SwitchButton(this.leftPos + 80, this.topPos + 8, "simplybackpacks.whitelist", ((this.menu.getFilterOpts() & 1) > 0) , (button)-> ((SwitchButton)button).state = (this.menu.saveFilterClient(this.menu.getFilterOpts() ^ 1) & 1) > 0));
        addRenderableWidget(new SwitchButton(this.leftPos + 80, this.topPos + 8 + 18, "simplybackpacks.nbtdata", ((this.menu.getFilterOpts() & 2) > 0) , (button)-> ((SwitchButton)button).state = (this.menu.saveFilterClient(this.menu.getFilterOpts() ^ 2) & 2) > 0));
        addRenderableWidget(new SwitchButton(this.leftPos + 80, this.topPos + 8 + 54, "simplybackpacks.autopickup", this.menu.getPickup() , (button)-> ((SwitchButton)button).state = this.menu.togglePickup()));

    }

    private final ResourceLocation GUI = ResourceLocation.fromNamespaceAndPath(SimplyBackpacks.MODID, "textures/gui/filter_gui.png");

    @Override
    public void render(@Nonnull GuiGraphics gg, int p_render_1_, int p_render_2_, float p_render_3_) {
        this.renderBackground(gg, p_render_1_, p_render_2_, p_render_3_);
        super.render(gg,p_render_1_, p_render_2_, p_render_3_);
        this.renderTooltip(gg, p_render_1_, p_render_2_);
    }

    @Override
    protected void renderBg(@Nonnull GuiGraphics gui, float partialTicks, int x, int y) {
        gui.blit(GUI, this.leftPos, this.topPos, 0,0, 176,166, 176, 166);
    }

    @Override
    protected void renderLabels(@Nonnull GuiGraphics gg, int x, int y) {
        // nope...
    }

    @Override
    protected void renderTooltip(@Nonnull GuiGraphics gg, int x, int y) {
        super.renderTooltip(gg, x, y);

        children().forEach((child) -> {
            if (child.isMouseOver(x,y) && child instanceof SlotButton button)
                if (!this.menu.filterHandler.getStackInSlot(button.slot).isEmpty())
                    gg.renderTooltip(font, this.menu.filterHandler.getStackInSlot(button.slot), x, y);
        });
    }

    class SlotButton extends Button {
        public SlotButton(int x, int y, int width, int height, int slotIn, OnPress pressable) {
            super(x,y,width,height,Component.empty(), pressable, Button.DEFAULT_NARRATION);

            this.slot = slotIn;
        }
        public final int slot;

        @Override
        public void renderWidget(@Nonnull GuiGraphics gg, int mouseX, int mouseY, float partialTicks) {
            boolean hovered = mouseX >= this.getX() && mouseX < this.getX() + this.width && mouseY >= this.getY() && mouseY < this.getY() + this.height;

            if (menu.filterHandler != null && !menu.filterHandler.getStackInSlot(this.slot).isEmpty()) {
                ItemStack tmp = menu.filterHandler.getStackInSlot(this.slot);
                RenderSystem.enableDepthTest();
                Lighting.setupForFlatItems();

                gg.renderItem(tmp, this.getX(), this.getY());

                Lighting.setupFor3DItems();
                RenderSystem.disableDepthTest();
            }

            if (hovered)
                gg.fill(this.getX(), this.getY(), this.getX() + this.width-1, this.getY() + this.height-1, -2130706433);
        }
    }


    class SwitchButton extends Button {
        public SwitchButton(int x, int y, String text, boolean initial, OnPress pressable) {
            super(x,y,32,16, Component.empty(), pressable, Button.DEFAULT_NARRATION);
            this.textKey = text;
            this.state = initial;
        }

        private static final ResourceLocation off = ResourceLocation.fromNamespaceAndPath(SimplyBackpacks.MODID, "textures/gui/switch_off.png");
        private static final ResourceLocation on = ResourceLocation.fromNamespaceAndPath(SimplyBackpacks.MODID, "textures/gui/switch_on.png");
        public boolean state;
        private final String textKey;

        @Override
        public void renderWidget(@Nonnull GuiGraphics gg, int p_renderButton_1_, int p_renderButton_2_, float p_renderButton_3_) {
            gg.blit(this.state ? on:off, this.getX(), this.getY(), this.width, this.height,0,0,32,16, 32 ,16);
            gg.drawString(font, I18n.get(this.textKey), this.getX() + 34, this.getY() + 4, 0x404040, false);
        }
    }
}

File: src/main/kotlin/com/flanks255/simplybackpacks/gui/FilterGui.kt
package com.flanks255.simplybackpacks.gui

import com.flanks255.simplybackpacks.simplybackpacks
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.GuiButton
import net.minecraft.client.gui.GuiButtonToggle
import net.minecraft.client.gui.inventory.GuiContainer
import net.minecraft.client.renderer.*
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.client.resources.I18n
import net.minecraft.util.ResourceLocation
import net.minecraftforge.items.IItemHandler
import org.lwjgl.opengl.GL11

class FilterGui(val container: FilterContainer): GuiContainer(container) {

    private val background: ResourceLocation = ResourceLocation(simplybackpacks.MODID, "textures/gui/filter_gui.png")
    private var filter: Int = 0

    init {
        xSize = 176
        ySize = 166
    }

    override fun initGui() {
        super.initGui()
        var slot = 0
        for (row in 0 until 4) {
            for (col in 0 until 4) {
                val x = guiLeft + 7 + col * 18
                val y = guiTop + 7 + row * 18
                buttonList.add(slot, slotButton(slot, x + 1, y + 1, slot))
                slot++
            }
        }

        buttonList.add(16, switchButton(16,guiLeft + 80, guiTop + 8, "simplybackpacks.whitelist"))
        buttonList.add(17, switchButton(17,guiLeft + 80, guiTop + 18 + 8, "simplybackpacks.metadata"))
        buttonList.add(18, switchButton(18,guiLeft + 80, guiTop + 36 + 8, "simplybackpacks.nbtdata"))

        updateSwitches()
    }

    fun drawTexturedQuadFit(x: Int, y: Int, width: Int, height: Int, z: Float) {
        val tessellator: Tessellator = Tessellator.getInstance()
        val buffer: BufferBuilder = tessellator.buffer

        buffer.begin(7, DefaultVertexFormats.POSITION_TEX)
        buffer.pos((x + 0).toDouble(), (y + height).toDouble(), z.toDouble()).tex(0.0,1.0).endVertex()
        buffer.pos((x + width).toDouble(), (y + height).toDouble(), z.toDouble()).tex(1.0,1.0).endVertex()
        buffer.pos((x + width).toDouble(), (y + 0).toDouble(), z.toDouble()).tex(1.0,0.0).endVertex()
        buffer.pos((x + 0).toDouble(), (y + 0).toDouble(), z.toDouble()).tex(0.0,0.0).endVertex()

        tessellator.draw()
    }

    override fun actionPerformed(button: GuiButton) {
        if (button is slotButton) {
            if (container.enchantItem(container.player, button.slot))
            mc.playerController.sendEnchantPacket(container.windowId, button.slot)
        }

        if (button is switchButton) {
            var opts = container.getFilterOpts()
            when(button.id) {
                16-> opts = opts xor 1
                17-> opts = opts xor 2
                18-> opts = opts xor 4
                else -> return
            }
            container.setFilterOpts(opts)
        }
        updateSwitches()
    }

    override fun drawGuiContainerBackgroundLayer(partialTicks: Float, mouseX: Int, mouseY: Int) {
        mc.textureManager.bindTexture(background)
        drawTexturedQuadFit(guiLeft, guiTop, xSize,ySize, this.zLevel)
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        this.drawDefaultBackground()
        super.drawScreen(mouseX, mouseY, partialTicks)
        this.renderHoveredToolTip(mouseX,mouseY)
    }

    override fun renderHoveredToolTip(x: Int, y: Int) {
        super.renderHoveredToolTip(x, y)

        for (button in buttonList) {
            if (button.isMouseOver && button is slotButton) {
                if (container.filterHandler?.getStackInSlot(button.slot)?.isEmpty == false) {
                    renderToolTip(container?.filterHandler?.getStackInSlot(button.slot), x, y)
                }
            }
        }
    }

    fun updateSwitches() {
        val opts = container.getFilterOpts()
        (buttonList[16] as switchButton).state = opts and 1 > 0
        (buttonList[17] as switchButton).state = opts and 2 > 0
        (buttonList[18] as switchButton).state = opts and 4 > 0
    }

    inner class slotButton(id: Int, x: Int, y: Int, val slot: Int): GuiButton(id, x, y, 16,16, "") {
        override fun drawButton(mc: Minecraft, mouseX: Int, mouseY: Int, partialTicks: Float) {
            if (visible) {
                GL11.glPushMatrix()
                GlStateManager.color(1F, 1F, 1F)
                hovered = mouseX >= x && mouseX < x + width && mouseY >= y && mouseY < y + height

                GlStateManager.enableBlend()
                GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO)
                GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA)

                if (hovered)
                    drawRect(x, y, x + width, y + height, -2130706433)

                if (container.filterHandler != null && container.filterHandler?.getStackInSlot(slot)?.isEmpty == false) {
                    zLevel = 100F
                    itemRender.zLevel = 100F
                    GlStateManager.enableDepth()
                    RenderHelper.enableGUIStandardItemLighting()
                    itemRender.renderItemAndEffectIntoGUI(mc.player, (container.filterHandler as IItemHandler).getStackInSlot(slot), x, y)
                    itemRender.renderItemOverlayIntoGUI(fontRenderer, (container.filterHandler as IItemHandler).getStackInSlot(slot), x, y, "")
                    zLevel = 0F
                    itemRender.zLevel = 0F
                }
                GL11.glPopMatrix()
                mouseDragged(mc, mouseX, mouseY)
            }
        }
    }

    inner class switchButton(id: Int, x: Int, y: Int, val textKey: String): GuiButton(id, x, y, 32, 16, "") {
        var state: Boolean = false
        private val off: ResourceLocation = ResourceLocation(simplybackpacks.MODID, "textures/gui/switch_off.png")
        private val on: ResourceLocation = ResourceLocation(simplybackpacks.MODID, "textures/gui/switch_on.png")

        fun drawTexturedQuadFit(x: Int, y: Int, width: Int, height: Int, z: Float) {
            val tessellator: Tessellator = Tessellator.getInstance()
            val buffer: BufferBuilder = tessellator.buffer
            GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F)
            buffer.begin(7, DefaultVertexFormats.POSITION_TEX)
            buffer.pos((x + 0).toDouble(), (y + height).toDouble(), z.toDouble()).tex(0.0,1.0).endVertex()
            buffer.pos((x + width).toDouble(), (y + height).toDouble(), z.toDouble()).tex(1.0,1.0).endVertex()
            buffer.pos((x + width).toDouble(), (y + 0).toDouble(), z.toDouble()).tex(1.0,0.0).endVertex()
            buffer.pos((x + 0).toDouble(), (y + 0).toDouble(), z.toDouble()).tex(0.0,0.0).endVertex()

            tessellator.draw()
        }

        override fun drawButton(mc: Minecraft, mouseX: Int, mouseY: Int, partialTicks: Float) {

            mc.textureManager.bindTexture(if (state) on else off)
            drawTexturedQuadFit(x, y, 32, 16, 100F)
            GL11.glPushMatrix()
            GL11.glColor4f(0.25f,0.25f,0.25f,1.0f)
            fontRenderer.drawString(I18n.format(textKey), x + 34, y + 4, 0x404040)
            GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F)
            GL11.glPopMatrix()
            mouseDragged(mc, mouseX, mouseY)
        }
    }
}